### PR TITLE
bpo-21302: Add clock_nanosleep() implementation for time.sleep()

### PR DESCRIPTION
**Pull request by Livius90**
- Add clock_nanosleep() function to time.sleep() in Unix operation systems.

Requested changes from @vstinner are done. It was runt-time tested in Oracle Linux 8.4 and Windows 10.
Some previous pull request history can be found here: https://github.com/python/cpython/pull/28077

https://bugs.python.org/issue21302

### CODE REVIEW

File: configure.ac [hunk 1]
```diff
@@ -4086,6 +4086,12 @@ AC_CHECK_FUNCS(clock_gettime, [], [
     ])
 ])
 
+AC_CHECK_FUNCS(clock_nanosleep, [], [
+    AC_CHECK_LIB(rt, clock_nanosleep, [
+        AC_DEFINE(HAVE_CLOCK_NANOSLEEP, 1)
+    ])
+])
```
Comment (vstinner) [inline configure.ac]: Would you mind to move it after the clock_settime check?

File: Modules/timemodule.c [hunk 2]
```diff
@@ -2066,14 +2070,24 @@ pysleep(_PyTime_t secs)
         return -1;
     }
     deadline = monotonic + secs;
+#ifdef HAVE_CLOCK_NANOSLEEP
+    if (_PyTime_AsTimespec(deadline, &timeout_abs) < 0)
```
Comment (vstinner) [inline Modules/timemodule.c]: Please add { ... } (PEP 7). Please also add them to the existing _PyTime_AsTimeval() call.

**Review by vstinner — COMMENTED** [review]
Oh nice, it's much better!

IMO it's worth to mention that time.sleep() has a resolution of 1 nanosecond if the clock_nanosleep() function is available.

Do you know if clock_nanosleep() is available on other platforms than Linux? Is it available on macOS or FreeBSD for example?

**Comment by Livius90** [comment]
> Oh nice, it's much better!
> 
> IMO it's worth to mention that time.sleep() has a resolution of 1 nanosecond if the clock_nanosleep() function is available.
> 
> Do you know if clock_nanosleep() is available on other platforms than Linux? Is it available on macOS or FreeBSD for example?

I do not think that it is available in macOS and FreeBSD according the [gnulib manual](https://www.gnu.org/software/gnulib/manual/html_node/clock_005fnanosleep.html#clock_005fnanosleep), clock_nanosleep() is available sure in Linux from POSIX version 2001.12. So any modern and new Linux based OS has it, moreover it seems to me [QNX has also the same clock_nanosleep()](http://www.qnx.com/developers/docs/6.5.0SP1.update/index.html#./com.qnx.doc.neutrino_lib_ref/c/clock_nanosleep.html).

Is there any right place to mention this improved resolution of time.sleep()?

**Comment by Livius90** [comment]
In all Unix systems eg: macOS, FreeBSD, Linux etc, [nanosleep()](https://www.gnu.org/software/gnulib/manual/html_node/nanosleep.html#nanosleep) is available and as i see there are some codes where select() is used for sleeping microsecs resolution, [semaphore.c](https://github.com/python/cpython/blob/main/Modules/_multiprocessing/semaphore.c#L276), [_tkinter.c](https://github.com/python/cpython/blob/main/Modules/_tkinter.c#L364). Sleeping with select() was the state of the art solution **in about 1998** in Linux, it is an **outdated solution nowadays**. It would be better to use nanosleep() in these codes also.

What is the rules to make a new pull request without bpo issue number? I would like to improve it in the rest of codes, too.

**Comment by Livius90** [comment]
Improved time.sleep() is ready for final review. All platforms have now the state of the art solution for sleeping. In Windows it uses Waitable timer object, in Linux it is clock_nanosleep() and in other Unix systems eg: macOS, FreeBSD it is nanosleep() function.

Resolution of Waitable timer object is 100 nsec but it is limited to 1 usec by round ceiling. Moreover sleep for lower then 1 millisec is not possible in Win32 API.

@abalkin, @pganssle and @vstinner please make review for final my state.

**Comment by Livius90** [comment]
I did some performance test and i compared the new time.sleep() implementations to old python3 versions. Following python code was my test script:

```python
import sys
import time
import statistics

SEC_TO_USEC = 1000000.0

def test_sleep(sleep_time, N):
    dt_array = [0.0]*N
    sleep_time_us = sleep_time*SEC_TO_USEC
    print("sleep [usec]: ", sleep_time_us)
    print("loop count: ", N, "\n")
    
    for i in range(N):
        t0 = time.perf_counter()
        time.sleep(sleep_time)
        dt_array[i] = time.perf_counter() - t0
    
    sleep_mean_us = statistics.mean(dt_array)*SEC_TO_USEC
    sleep_stdev_us = statistics.stdev(dt_array)*SEC_TO_USEC
    sleep_diff = sleep_mean_us - sleep_time_us
    print("sleep mean [usec]: ", sleep_mean_us)
    print("sleep stdev [usec]: ", sleep_stdev_us)
    print("sleep diff [usec]: ", sleep_diff)
    
if __name__ == "__main__":
    print("Python version:", sys.version)
    print("\n")
    test_sleep(0.2, 2048)
```

Test result can be here for Linux. It was executed on an Oracle Linux 8.4 :
```console
[livius@test-pc cpython]$ python3 ../test_sleep.py
Python version: 3.6.8 (default, May 19 2021, 10:00:09)
[GCC 8.4.1 20200928 (Red Hat 8.4.1-1.0.1)]


sleep [usec]:  200000.0
loop count:  2048

sleep mean [usec]:  200284.74022315067
sleep stdev [usec]:  24.013486769335735
sleep diff [usec]:  284.7402231506712


[livius@test-pc cpython]$ ./python ../test_sleep.py
Python version: 3.11.0a0 (heads/bpo-21302:db7851bf47, Sep  4 2021, 14:57:19) [GCC 8.4.1 20200928 (Red Hat 8.4.1-1.0.3)]


sleep [usec]:  200000.0
loop count:  2048

sleep mean [usec]:  200138.9896440342
sleep stdev [usec]:  10.762552111307311
sleep diff [usec]:  138.9896440342127
```

Test result can be here for Windows 10:
```console
C:\Work>python.exe test_sleep.py
Python version: 3.7.8 (tags/v3.7.8:4b47a5b6ba, Jun 28 2020, 08:53:46) [MSC v.1916 64 bit (AMD64)]


sleep [usec]:  200000.0
loop count:  2048

sleep mean [usec]:  204847.39228515618
sleep stdev [usec]:  3421.194947272968
sleep diff [usec]:  4847.39228515618


C:\Work\cpython>python.bat ../test_sleep.py
Running Release|x64 interpreter...
Python version: 3.11.0a0 (heads/bpo-21302:db7851bf47, Sep  4 2021, 15:41:23) [MSC v.1929 64 bit (AMD64)]


sleep [usec]:  200000.0
loop count:  2048

sleep mean [usec]:  204558.17597656834
sleep stdev [usec]:  3207.7529824030335
sleep diff [usec]:  4558.175976568338
```

File: Modules/_tkinter.c [hunk 3]
```diff
@@ -357,11 +357,10 @@ static int Tkinter_busywaitinterval = 20;
...
 Sleep(int milli)
 {
-    /* XXX Too bad if you don't have select(). */
-    struct timeval t;
+    struct timespec t;
     t.tv_sec = milli/1000;
-    t.tv_usec = (milli%1000) * 1000;
-    select(0, (fd_set *)0, (fd_set *)0, (fd_set *)0, &t);
+    t.tv_nsec = (milli%1000) * 1000000;
+    nanosleep(&t, NULL);
```
Comment (vstinner) [inline Modules/_tkinter.c]: Please revert this change and write a second PR for this change.

File: Misc/NEWS.d/next/Library/2021-09-04-11-46-05.bpo-21302.zTYR1A.rst [hunk 4]
```diff
@@ -0,0 +1 @@
+In Unix operating systems eg: Linux, macOS, FreeBSD etc. time.sleep() has a resolution of nanoseconds with using clock_nanosleep() or nanosleep() function. In Windows, time.sleep() has a resolution of microseconds with using waitable timer from Win32 API.
```
Comment (vstinner) [inline Misc/NEWS.d/next/Library/2021-09-04-11-46-05.bpo-21302.zTYR1A.rst]: The Windosw code path is not modified, so you should not mention it in the *Change* log.

**Review by vstinner — COMMENTED** [review]
I would prefer to keep the select() code path since this one has been proved to work on all platforms (but Windows). If you want to use nanosleep(), I would prefer to add a configure check to check if the function is available.

Is it important for you to use nanosleep() on platforms where clock_nanosleep() is not available?

Maybe I'm wrong and nanosleep() is available "everywhere".

**Comment by Livius90** [comment]
> I would prefer to keep the select() code path since this one has been proved to work on all platforms (but Windows). If you want to use nanosleep(), I would prefer to add a configure check to check if the function is available.
> 
> Is it important for you to use nanosleep() on platforms where clock_nanosleep() is not available?
> 
> Maybe I'm wrong and nanosleep() is available "everywhere".

- [nanosleep()](https://www.gnu.org/software/gnulib/manual/html_node/nanosleep.html#nanosleep) is supported in the following operating systems: [Linux (since 2001)](https://man7.org/linux/man-pages/man2/nanosleep.2.html), [FreeBSD (since 1997)](https://www.freebsd.org/cgi/man.cgi?query=nanosleep&apropos=0&sektion=0&manpath=FreeBSD+14.0-current&arch=default&format=html), [macOS (since 2013)](https://developer.apple.com/library/archive/documentation/System/Conceptual/ManPages_iPhoneOS/man2/nanosleep.2.html)
- clock_nanosleep() is supported in the following operating systems: [Linux (since 2001)](https://man7.org/linux/man-pages/man2/clock_nanosleep.2.html), [FreeBSD (since 2008)](https://wiki.freebsd.org/FreeBSD_and_Standards)

What more unix operating systems need to be supported for Python 3?

Please look at my new Windows implementation of time.sleep(), too. I changed the sleep functionality, it is based on a waitable timer object instead of a dummy singal waiting and timeout.

File: Modules/_tkinter.c (hunk 3 above)
Comment (Livius90) [reply Modules/_tkinter.c]: Luckily nanosleep() is supported in many unix OS, see my latest comment.
https://github.com/python/cpython/pull/28111#issuecomment-914642746

Do you suggest to do it in a new PR for changes of [semaphore.c](https://github.com/python/cpython/pull/28111/files#diff-cf698f2997f9f11b2851514321974536c9a161585e85034230a9783c4a7bc98e), also?

File: Misc/NEWS.d/next/Library/2021-09-04-11-46-05.bpo-21302.zTYR1A.rst (hunk 4 above)
Comment (Livius90) [reply Misc/NEWS.d/next/Library/2021-09-04-11-46-05.bpo-21302.zTYR1A.rst]: Please look at my new Windows implementation of time.sleep(), too. I changed the sleep implementation in Windows, it is based on a Win32 waitable timer object instead of a dummy signal waiting and timeout.

**Comment by vstinner** [comment]
Sorry, but this PR combines too many changes:
* use clock_nanosleep()
* use nanosleep()
* change the time.sleep() implementation on Windows

Can you please start with a PR which only starts using clock_nanosleep() with the configure check? Then maybe change the Windows implementation, to finish with a 3rd PR for nanosleep().

> nanosleep() is supported in the following operating systems: Linux (since 2001), FreeBSD (since 1997), OpenBSD (since 1997), NetBSD (since 1997), IBM AIX (since v7.1 2010), macOS (since OS X v10.9 2013)

If I recall correctly, Python targets macOS 10.6, according to my notes: https://pythondev.readthedocs.io/platforms.html#macos

**Comment by Livius90** [comment]
PR was changed to add clock_nanosleep() only. Can you review it?

File: Misc/NEWS.d/next/Library/2021-09-10-00-24-08.bpo-21302.XS_frc.rst [hunk 5]
```diff
@@ -0,0 +1 @@
+In Unix operating systems time.sleep() has a resolution of nanoseconds with using clock_nanosleep() function.
```
Comment (vstinner) [inline Misc/NEWS.d/next/Library/2021-09-10-00-24-08.bpo-21302.XS_frc.rst]:
```suggestion
In Unix operating systems, :func:`time.sleep` now uses the ``clock_nanosleep()`` function,
if available, which allows to sleep for an interval specified with nanosecond precision.
```

File: Modules/timemodule.c [hunk 6]
```diff
@@ -2053,7 +2053,11 @@ pysleep(_PyTime_t secs)
 {
     _PyTime_t deadline, monotonic;
 #ifndef MS_WINDOWS
+#ifdef HAVE_CLOCK_NANOSLEEP
+    struct timespec timeout;
```
Comment (vstinner) [inline Modules/timemodule.c]: Usually, a timeout is relative. This one is special, it's absolute. Can you please rename it to "timeout_abs"?

File: Modules/timemodule.c [hunk 7]
```diff
@@ -2066,23 +2070,44 @@ pysleep(_PyTime_t secs)
...
         Py_END_ALLOW_THREADS
 
-        if (err == 0)
+        if (err == 0) {
             break;
+        }
 
+#ifdef HAVE_CLOCK_NANOSLEEP
+        if (err != EINTR) {
+            errno = err;
```
Comment (vstinner) [inline Modules/timemodule.c:2105]: Can you merge the two if/else code  paths? Only this line is different if HAVE_CLOCK_NANOSLEEP is defined.

File: Modules/timemodule.c [hunk 8]
```diff
@@ -2114,8 +2139,9 @@ pysleep(_PyTime_t secs)
 #endif
 
         /* sleep was interrupted by SIGINT */
-        if (PyErr_CheckSignals())
+        if (PyErr_CheckSignals()) {
             return -1;
+        }
 
         if (get_monotonic(&monotonic) < 0) {
```
Comment (vstinner) [inline Modules/timemodule.c:2145]: Recomputing secs is useless if clock_nanosleep() is used no? It might be more efficient to avoid the get_monotonic() syscall.

Comment (Livius90) [reply Modules/timemodule.c:2145]: Yes, recomputing of monotonic time is totally useless if clock_nanosleep() is used.

File: Modules/timemodule.c (hunk 7 above)
Comment (Livius90) [reply Modules/timemodule.c:2105]: In case of using select(), errno must be compared in if statement. If clock_nanosleep() is used its (err) return value must be in the if statement comparing. If i am not following this, some unit tests will be failed because there are any magic behavier/differences in errno and the returned err of clock_nanosleep() and select(). I can not merge them better.

File: Modules/timemodule.c (hunk 6 above)
Comment (Livius90) [reply Modules/timemodule.c]: Done

File: Modules/timemodule.c (hunk 7 above)
Comment (Livius90) [reply Modules/timemodule.c:2105]: I put errno value passing between separated Py_BEGIN_ALLOW_THREADS and Py_END_ALLOW_THREADS.

**Comment by Livius90** [comment]
@vstinner 
"Tests / Address sanitizer (pull_request)" is randomly failed or succes for same code, can you re-run it somehow?

**Comment by vstinner** [comment]
Merged, thanks! I created PR #28311 follow-up to update the documentation.

**Comment by Livius90** [comment]
Thanks the merging.

What are the formal rules to create a new PR to improve Windows implementation, too? Is it good to create a new PR without any bpo issue number? Or can i make more PR (windows implementation and nanosleep()) for this bpo-21302?

**Comment by vstinner** [comment]
In this case, you can reuse bpo-21302.
